zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/archiver/OpenSearchRepositoryIT.java

```diff
@@ -81,6 +87,81 @@ void shouldDeleteDocuments() throws IOException {
         .isEqualTo("3");
   }
 
+  @Test
+  void shouldSetIndexLifeCycle() throws IOException {
+    // given
+    final var indexName = UUID.randomUUID().toString();
+    final var repository = createRepository();
+    testClient.indices().create(r -> r.index(indexName));
+
+    final OpensearchEngineClient engineClient = new OpensearchEngineClient(testClient);
+    final OpenSearchGenericClient genericClient =
+        new OpenSearchGenericClient(testClient._transport(), testClient._transportOptions());
+
+    retention.setEnabled(true);
+    retention.setPolicyName("operate_delete_archived_indices");
+
+    // when
+    createLifeCyclePolicy(engineClient);
+    final var result = repository.setIndexLifeCycle(indexName);
+
+    // then
+    assertThat(result).succeedsWithin(Duration.ofSeconds(30));
+
+    // Takes a while for the policy to be applied
+    Awaitility.await()
+        .atMost(3, TimeUnit.SECONDS)
+        .pollInterval(1, TimeUnit.SECONDS)
+        .until(() -> indexContainsRetentionPolicy(indexName, genericClient));
```

Comment:
:x: I would change the assertion. What happens now when it fails? You just get that you expected something to be true, but it was false, right? It doesn't give you a lot of information on failure, which will make it difficult when you want to diagnose problems.

Instead, we can try and get some of that information in the failure message.

Let's change the signature of `indexContainsRetentionPolicy`, and instead have a method which just returns the policy of the index. For that, we first need to be able to parse the body. Since the sample response from the API is:

```json
{
  "index_1": {
    "index.plugins.index_state_management.policy_id": "policy_1"
  }
}
```

Introducing a type to parse that would be difficult, because the top level field names are basically the index name - so they're dynamic! So it's hard to define a POJO for deserialization. But since the structure is simple enough, we can just use the basic JSON API.

We add a method which fetches the policy for a given index. I saw you used the asynchronous client, but it's much easier and less flaky in a test to just use the synchronous variant:

> [!Note]
> We'd have created a static `ObjectMapper MAPPER` property in the class before.

```java
  private String fetchPolicyForIndex(
      final String indexName, final OpenSearchGenericClient genericClient) {
    final var request =
        Requests.builder().method("get").endpoint("_plugins/_ism/explain/" + indexName).build();
    try {
      final var response = genericClient.execute(request);
      final var jsonString = response.getBody().orElseThrow().bodyAsString();
      final var json = MAPPER.readTree(jsonString);
      final var index = json.get(indexName);
      if (index == null) {
        throw new AssertionError(
            "Failed to explain non-existent index '%s'; see response: %s"
                .formatted(indexName, jsonString));
      }

      return index.findPath("index.plugins.index_state_management.policy_id").asText();
    } catch (final IOException e) {
      throw new AssertionError("Failed to fetch policy for index " + indexName, e);
    }
  }
```

Then we can use the `untilAsserted` variant of `Awaitility`, which will print out the last failed assertion on failure, giving us some more information about why the test failed.

```java
    Awaitility.await("until the policy has been visibly applied")
        .untilAsserted(
            () ->
                assertThat(fetchPolicyForIndex(indexName, genericClient))
                    .isEqualTo(retention.getPolicyName()));
```

> [!Note]
> We also don't really need to override defaults of Awaitility - let's do that only if we see a need to after it's already running for a while.

It's really important to have good error messages on test failures, otherwise it's easy to lose time trying to understand what went wrong.

Reply:
Good point, the previous error message was not useful at all, thanks for the suggestion 🙌